services/libs/aster-std/src/sched/nice.rs

```diff
@@ -0,0 +1,141 @@
+
+/// The process scheduling priority value.
+///
+/// It is a value in the range -20 to 19, with -20 being the highest priority
+/// and 19 being the lowest priority. The smaller values give a process a higher
+/// scheduling priority.
+pub struct Nice {
+    // Use `u8` with the range [0, 39] to represent the real range [-20, 19]
+    value: AtomicU8,
+}
+
+impl Nice {
+    /// The minimum nice, whose value is -20.
+    pub fn min_value() -> Self {
+        Self {
+            value: AtomicU8::new(0),
+        }
+    }
+
+    /// The maximum nice, whose value is 19.
+    pub fn max_value() -> Self {
+        Self {
+            value: AtomicU8::new(39),
+        }
+    }
+
+    /// Returns the raw value with in range [-20, 19]
+    pub fn as_i32(&self) -> i32 {
+        (self.value.load(Ordering::Relaxed) as i32) - 20
+    }
+
+    /// Updates the nice value with `new_value`.
+    pub fn update(&self, new_value: &Nice) {
+        let new_value = new_value.value.load(Ordering::Relaxed);
+        self.value.store(new_value, Ordering::Relaxed);
+    }
+}
+
+impl Debug for Nice {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        f.debug_struct("Nice")
+            .field("value", &self.as_i32())
+            .finish()
+    }
+}
+
+impl Default for Nice {
+    fn default() -> Self {
+        Self {
+            // The default nice value is 0.
+            value: AtomicU8::new(20),
+        }
+    }
+}
+
+impl Clone for Nice {
+    fn clone(&self) -> Self {
+        let value = self.value.load(Ordering::Relaxed);
+        Self {
+            value: AtomicU8::new(value),
+        }
+    }
+}
+
+impl PartialEq for Nice {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_i32() == other.as_i32()
+    }
+}
+
+impl Eq for Nice {}
+
+impl PartialOrd for Nice {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Nice {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        self.as_i32().cmp(&other.as_i32())
+    }
+}
+
+impl From<i32> for Nice {
+    fn from(raw: i32) -> Self {
+        if raw <= Self::min_value().as_i32() {
+            Self::min_value()
+        } else if raw >= Self::max_value().as_i32() {
+            Self::max_value()
+        } else {
+            Self {
+                value: AtomicU8::new((raw + 20) as u8),
+            }
+        }
+    }
+}
+
+impl From<RlimitNice> for Nice {
+    fn from(rlimit_nice: RlimitNice) -> Self {
+        let raw: i32 = 20 - rlimit_nice.to_raw() as i32;
+        Self {
+            value: AtomicU8::new((raw + 20) as u8),
+        }
+    }
+}
+
+/// The process scheduling priority value used in rlimit.
+///
+/// It is a value in the range 40 to 1, with 40 being the highest priority
+/// and 1 being the lowest priority. The greater values give a process a higher
+/// scheduling priority.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RlimitNice {
```

Comment:
Call this `Priority` and explain its relationship with `Nice`.

Reply:
Done